server/src/main/java/org/elasticsearch/indices/recovery/RecoveryTarget.java

```diff
@@ -284,11 +285,21 @@ private void ensureRefCount() {
     /*** Implementation of {@link RecoveryTargetHandler } */
 
     @Override
-    public void prepareForTranslogOperations(boolean fileBasedRecovery, int totalTranslogOps, ActionListener<Void> listener) {
+    public void prepareForTranslogOperations(boolean fileBasedRecovery, int totalTranslogOps, long recoverUpToSeqNo,
+                                             ActionListener<Optional<Store.MetadataSnapshot>> listener) {
         ActionListener.completeWith(listener, () -> {
-            state().getTranslog().totalOperations(totalTranslogOps);
-            indexShard().openEngineAndSkipTranslogRecovery();
-            return null;
+            if (Assertions.ENABLED && recoverUpToSeqNo != SequenceNumbers.UNASSIGNED_SEQ_NO) {
+                final long gcp = readGlobalCheckpointFromTranslog();
+                assert recoverUpToSeqNo == gcp : recoverUpToSeqNo + " != " + gcp;
+            }
+            indexShard().openEngineAndRecoverFromTranslog(recoverUpToSeqNo);
+            if (indexShard.getLocalCheckpoint() < recoverUpToSeqNo) {
+                // This can happen if the previous recovery was aborted after copying segments, then the target does not have all translog.
+                indexShard.closeEngine();
+                return Optional.of(indexShard.snapshotStoreMetadata());
```

Comment:
I need to add a comment here. When we open an internal engine, we may trim the last commit (if it's unsafe) which we sent in the start recovery request. If we don't update the latest commit to the recovery source, it can send an incorrect set of diff files in a file-based recovery.